### PR TITLE
feat(runtime-vapor): component emits

### LINKED ISSUE #41: Component Emits
related #4  

I've submitted a PR concerning `Props`, so I'll proceed with the implementation of `Emits` based on [that branch](https://github.com/vuejs/core-vapor/pull/40).

### PR DESCRIPTION
**Pull request by Ubugeeei**
Closes #41 

## Summary

I have implemented emit.
This branch is derived from the [branch for implementing props](https://github.com/vuejs/core-vapor/pull/40). 
Please merge [the props PR](https://github.com/vuejs/core-vapor/pull/40) first.

- main
  - https://github.com/vuejs/core-vapor/pull/40
    - https://github.com/vuejs/core-vapor/pull/42 <- current
        - https://github.com/vuejs/core-vapor/pull/47
            -  https://github.com/vuejs/core-vapor/pull/54
## Points I'm unsure about

1. Flags related to lifecycle
   I had implemented it so that isMounted is set to false when unmounting,
   In traditional components, it seems that isUnmounted is set to true.
   Which approach should we follow for the implementation? (In this PR, I have added the isUnmounted flag.)

2. Holding rawProps in ComponentInternalInstance
   I think we will face the same issue when implementing attr, but in emit, we need to search for handlers from user-defined raw props, not those defined as propsOption.
   In the past, raw props were held in instance.vnode.props, but at the moment, there is no implementation that handles vnode, so I am directly holding them in ComponentInternalInstance.
   Is this okay?

3. Caching
   Similar to props, I haven't implemented caching because I don't understand how to handle it.
   Should we also implement it in Vapor Mode?

4. Validations
  Should be emits validations are implemented?
  https://vuejs.org/guide/components/events.html#events-validation

## How to confirm the operation

Since the compiler implementation is not yet available, I have written the expected JavaScript code as the compilation result in `playground/src/emits.js`.

<details>
<summary>
  <h3>code</h3>
</summary>

```ts
import {
  children,
  on,
  ref,
  template,
  watchEffect,
  setText,
  render as renderComponent // TODO:
} from '@vue/vapor'

export default {
  props: undefined,

  setup(_, {}) {
    const count = ref(1)
    const setCount = v => {
      count.value = v
    }

    const __returned__ = { count, setCount }

    Object.defineProperty(__returned__, '__isScriptSetup', {
      enumerable: false,
      value: true
    })

    return __returned__
  },

  render(_ctx) {
    const t0 = template('<p></p>')
    const n0 = t0()
    const {
      0: [n1]
    } = children(n0)

    watchEffect(() => {
      setText(n1, void 0, _ctx.count)
    })

    renderComponent(
      child,
      {
        get count() {
          return _ctx.count
        },
        get ['onClick:child']() {
          return _ctx.setCount
        }
      },
      [],
      n0
    )

    renderComponent(
      child,
      {
        get count() {
          return _ctx.count
        },
        get ['onClick:childOnce']() {
          return _ctx.setCount
        }
      },
      [],
      n0
    )

    return n0
  }
}

const child = {
  props: {
    count: { type: Number, default: 1 }
  },

  setup(props, { emit }) {
    const handleClick = () => {
      emit('click:child', props.count * 2)
    }

    const __returned__ = { handleClick }

    Object.defineProperty(__returned__, '__isScriptSetup', {
      enumerable: false,
      value: true
    })

    return __returned__
  },

  render(_ctx) {
    const t0 = template('<button>set count * 2</button>')
    const n0 = t0()
    const {
      0: [n1]
    } = children(n0)
    on(n1, 'click', _ctx.handleClick)
    return n0
  }
}
```

</details>

Left button: **normal emit**
Right button: **emit once**

https://github.com/vuejs/core-vapor/assets/71201308/7f4f06a8-f49b-46ed-bb6c-c3cbf78677c3

### CODE REVIEW

**Comment by Ubugeeei** [comment]
⚠️  Mark as draft until [Props PR](https://github.com/vuejs/core-vapor/pull/40) is merged.

File: packages/runtime-vapor/src/componentEmits.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+  hyphenate,
+  isArray,
+  isFunction,
+  toHandlerKey,
+} from '@vue/shared'
+import { type Component, type ComponentInternalInstance } from './component'
+
+export type ObjectEmitsOptions = Record<
+  string,
+  ((...args: any[]) => any) | null // TODO: call validation?
```
Comment (Ubugeeei) [inline packages/runtime-vapor/src/componentEmits.ts:16]: Should be emits validations are implemented?

File: playground/src/props.js
Comment (Ubugeeei) [inline playground/src/props.js:1]: I've just added a playground for checking the default values.
It's not related to the content of this PR, though.

Comment (Ubugeeei) [reply playground/src/props.js:1]: https://github.com/vuejs/core-vapor/assets/71201308/c4d4b860-0bc9-4c9f-9ac9-fbc332bd9c87

File: playground/src/emits.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+      n0
+    )
+
+    renderComponent(
+      child,
+      {
+        get count() {
+          return _ctx.count
+        },
+        'onClick:childOnce': _ctx.setCount
```
Comment (Ubugeeei) [inline playground/src/emits.js:58]: It might be problematic not to use a getter.

Comment (Ubugeeei) [reply playground/src/emits.js]: fixed

**Comment by Ubugeeei** [comment]
https://github.com/vuejs/core-vapor/pull/47#issuecomment-1869042881

**Comment by Ubugeeei** [comment]
This PR is on hold until this issue is resolved.
https://github.com/vuejs/core-vapor/issues/84

**Comment by Ubugeeei** [comment]
🤷‍♂️

```
Run pnpm run format-check
  pnpm run format-check
  shell: /usr/bin/bash -e {0}
  env:
    PUPPETEER_SKIP_DOWNLOAD: true
    PNPM_HOME: /home/runner/setup-pnpm/node_modules/.bin

> @0.0.0-vapor format-check /home/runner/work/core-vapor/core-vapor
> prettier --check --cache .

Checking formatting...
[warn] packages/reactivity/src/ref.ts
[warn] packages/runtime-core/src/componentEmits.ts
[warn] packages/runtime-core/src/componentPublicInstance.ts
[warn] packages/runtime-vapor/src/componentEmits.ts
[warn] tsconfig.json
[warn] Code style issues found in [5](https://github.com/vuejs/core-vapor/actions/runs/7600943756/job/20699756554?pr=42#step:7:5) files. Run Prettier to fix.
```

**Comment by sxzz** [comment]
https://github.com/vuejs/core-vapor/pull/83#issuecomment-1913580670

**Pull request closed without merging**